// coldfusion/lib/session-factory.js

// Include core libraries.
var events = require( "events" );

// Include the evented cache constructor. The session is a 
// specialized cache.
var EventedCache = require( "./evented-cache" );


// ---------------------------------------------------------- //
// ---------------------------------------------------------- //


// I am a session constructor.
function Session( sessionID, timeout ){

	// Call the super constructor.
	EventedCache.call( this );
	
	var that = this;

	// Store the unique session ID.
	this._sessionID = sessionID
	
	// Store the time at which the session started.
	this._started = (new Date().getTime());
	
	// Store the time at which the session was last "touched".
	this._lastTouched = this._started;
	
	// I store the timeout for the session. This the number of
	// milliseoncds that the session can stay idle before it times out.
	this._timeout = (timeout || (20 * 60 * 1000));
	
	// I am the timeout timer. The session can only sit idle for so 
	// long before it times out.
	this._timeoutTimer = setTimeout(
		function(){
			that.emit( "timeout" );
		},
		this._timeout
	);

}


// Extend the event emitter class.
Session.prototype = Object.create( EventedCache.prototype );


// I return the session's unique ID.
Session.prototype.getSessionID = function(){

	// Return the unique ID. 
	return( this._sessionID );
	
};


// I touch the session, breaking its idle time. I can also accept a new 
// timeout value (as the session timeout can be changed at any time).
Session.prototype.touch = function( newTimeout ){
	
	var that = this;

	// Clear any timer.
	clearTimeout( this._timeoutTimer );

	// Store the last touch date.
	this._lastTouched = (new Date()).getTime();

	// Store the new timeout (if provided).
	this._timeout = (newTimeout || this._timeout);
	
	// Set the new timer for our idle timeout.
	this._timeoutTimer = setTimeout(
		function(){
			that.emit( "timeout" );
		},
		this._timeout
	);
	
	// Return this object reference for method chaining.
	return( this );
	
};


// ---------------------------------------------------------- //
// ---------------------------------------------------------- //


// I am the session factory constructor.
function SessionFactory(){

	// Call the super constructor.
	events.EventEmitter.call( this );
	
	var that = this;
	
	// I am an auto-incrementing ID. Each session identifier will
	// be based on this ID and a time-string.
	this._sessionIndex = 0;
	
	// I am the collection of active sessions. Each session will be
	// indexed by its sessionID. 
	this._sessions = {};
	
}


// Extend the event emitter class. This will allow us to use the 
// on() and emit() methods.
SessionFactory.prototype = Object.create( events.EventEmitter.prototype );


// I return a new unique session ID.
SessionFactory.prototype._getNewSessionID = function(){
		
	// Return an incremented / random session ID.
	return( "CFJS-" + ++this._sessionIndex + "-" + (new Date()).getTime() );
	
};

	
// I get the session based on the given ID.
SessionFactory.prototype.getSession = function( sessionID ){
	
	// Return the session. If it doesn't exist, this will return
	// null.
	return( this._sessions[ sessionID ] || null );
	
};
	
	
// I create and return a new session.
SessionFactory.prototype.newSession = function( timeout ){
	
	var that = this;
	
	// Create a new session instance.
	var session = new Session( this._getNewSessionID(), timeout );
	
	// Cache the session in the local collection.
	this._sessions[ session.getSessionID() ] = session;
	
	// Bind a one-time listener to the session for timeout. Once
	// the session times out once, the session factory does not
	// care about it any longer.
	session.once(
		"timeout",
		function(){

			// Delete the session from the active sessions.
			delete( that._sessions[ session.getSessionID() ] );
			
			// Emit the timeout event at the factory level and pass the
			// target session through.
			that.emit( "timeout", session );
			
		}
	);
	
	// Retuen the session instance.
	return( session );
	
};
	

// ---------------------------------------------------------- //
// ---------------------------------------------------------- //

// Create an instance of the session factory.
var sessionFactory = new SessionFactory();

// Export this session factory.
module.exports = sessionFactory;
